platforms/jvm/plugins-java/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -268,43 +277,47 @@ public void apply(final Project project) {
         configureSourceSets(javaExtension, buildOutputCleanupRegistry);
 
         configureTestTaskOrdering(project.getTasks());
-        configureDiagnostics(project, javaComponent);
+        configureDiagnostics(project, mainFeature);
         configureBuild(project);
     }
 
-    private static DefaultJvmSoftwareComponent createJavaComponent(Project project, JavaPluginExtension javaExtension) {
-        // Create the 'java' component - create sourceset first
-        SourceSet sourceSet = createSourceSet(SourceSet.MAIN_SOURCE_SET_NAME, javaExtension.getSourceSets());
+    private static JvmFeatureInternal createMainFeature(ProjectInternal project, JavaPluginExtension javaExtension) {
+        // Create the 'main' feature - create source set first
+        SourceSetContainer sourceSets = javaExtension.getSourceSets();
+        SourceSet sourceSet = sourceSets.create(SourceSet.MAIN_SOURCE_SET_NAME);
 
         // Supply the sourceSet to the feature
         JvmFeatureInternal feature = new DefaultJvmFeature(
-            JvmConstants.JAVA_MAIN_FEATURE_NAME, sourceSet, Collections.emptyList(),
-            (ProjectInternal) project, false, false);
+            JvmConstants.JAVA_MAIN_FEATURE_NAME,
+            sourceSet,
```

Comment:
❓ I thought you wanted features to create their source sets upon creation?

Reply:
Eventually, not in this PR though